The Big Guide to Web Components Built With React/The Big Guide to Web Components Built With React.md

```diff
@@ -0,0 +1,432 @@
+
+<iframe
+     src="https://codesandbox.io/embed/react-web-component-with-reacthtmlelement-7gsrj?fontsize=14&hidenavigation=1&initialpath=%2Fhome&theme=dark"
+     style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;"
+     title="React Web Component With ReactHTMLElement"
+     allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr"
+     sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
+   ></iframe>
+
+This was our first consideration, as it is the most clear problem when building Web Components with React. The next problem we wanted to solve was: how do we render HTML Children within a Web Component?
+
+## Rendering Externally-Provided Children
+
+The Web Components APIs provide [a very simple way of rendering children using `<slot>` tags](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_templates_and_slots#Adding_flexibility_with_slots). Luckily, we found that using `<slot>` in React was easy, and it just works. You can manage `<slot>` tags with `name` attributes, but the absolute simplest way is just to include a bare `<slot>` tag in your React component:
+
+```react
+export default function AppWithSlot() {
+  <div>
+    Here are my children:
+    <div>
+      <slot></slot>
+    </div>
+  </div>
+}
+```
+
+With that in place (and with your app registered as a Web Component called `react-app`), your HTML could look something like this:
+
+```html
+<html>
+  <head>
+    <!-- load your js, Web Components polyfills, etc -->
+  </head>
+  <body>
+    <react-app>
+      <div>
+        Any children can go in here, including other Web Components, and even other Web Components built with React.
+      </div>
+    </react-app>
+  </body>
+</html>
+```
+
+That's it! A very simple solution built right into the browser!
+
+The next question we wanted to answer was: how do Web Components communicate with each other?
+
+## Inter-Component Communication
+
+There are really two ways to handle this problem. You can  [build a Web Component with a JavaScript API](https://developers.google.com/web/fundamentals/web-components/customelements#jsapi), and pass data into the Web Component using a reference to the element itself by calling the Javascript API you have defined. This method is good, but probably the simplest way for Web Components built with React is to use events on the Window. An app can emit a Window event, and another can listen for it to update based on any information that has changed.
```

Comment:
Remind me why we went with using events on the window instead of the web component's root?

Reply:
All components have access to the `window` object. To listen for an event on another component’s root, you’d have to know about the other component from yours, or your component would have to ensure that you set `bubbles` to `true` to make sure that a component listening in the window would catch it.

It seems easier just for every Web Component to dispatch and listen for events directly on the window, and then make sure that our event names are clear so we’re not colliding.